macros/generic/optional_table.sql

```diff
@@ -0,0 +1,25 @@
+{%- macro optional_table(source_table) -%}
+
+{%- set source_relation = adapter.get_relation(
+    database=source_table.database,
+    schema=source_table.schema,
+    identifier=source_table.name) -%}
+
+{# When source table does not exist, create an empty table in the target schema that will be queried instead. #}
+{%- if source_relation is not none -%}
+    {{ source_table }}
+{%- else -%}
+    {% set query %}
+        {% if target.type == 'snowflake' %}
+            create or replace table "{{ target.database }}"."{{ target.schema }}"."{{ source_table.name }}" (dummy int)
+        {% elif target.type == 'sqlserver' %}
+            if object_id('"{{ target.database }}"."{{ target.schema }}"."{{ source_table.name }}"', 'U') is null
```

Comment:
Why is this extra check necessary? It does not break anything, but the check for existence of the table already happened in the test for the source_relation?

Reply:
- This check is necessary to achieve the behavior of "create or replace". We can not create the dummy table if it is already created. Without this check, it breaks during development if your run for a second time your dbt project.
- It is a different check because we do not check whether the table exists in the source schema, but in the target schema (= the dummy table).